File: clients/typescript/src/model/vec2.ts
import { Stream } from "../stream";

/**
 * 2 dimensional vector.
 */
export class Vec2 {
    /**
     * `x` coordinate of the vector
     */
    x: number
    /**
     * `y` coordinate of the vector
     */
    y: number

    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }

    /**
     * Read Vec2 from input stream
     */
    static async readFrom(stream: Stream): Promise<Vec2> {
        let x;
        x = await stream.readDouble();
        let y;
        y = await stream.readDouble();
        return new Vec2(x, y)
    }

    /**
     * Write Vec2 to output stream
     */
    async writeTo(stream: Stream) {
        let x = this.x;
        await stream.writeDouble(x);
        let y = this.y;
        await stream.writeDouble(y);
    }
}

File: clients/python/model/vec2.py
from stream_wrapper import StreamWrapper

class Vec2:
    """2 dimensional vector."""

    __slots__ = ("x","y",)

    x: float
    y: float

    def __init__(self, x: float, y: float):
        self.x = x
        """`x` coordinate of the vector"""
        self.y = y
        """`y` coordinate of the vector"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Vec2":
        """Read Vec2 from input stream
        """
        x = stream.read_double()
        y = stream.read_double()
        return Vec2(x, y)
    
    def write_to(self, stream: StreamWrapper):
        """Write Vec2 to output stream
        """
        stream.write_double(self.x)
        stream.write_double(self.y)
    
    def __repr__(self):
        return "Vec2(" + \
            repr(self.x) + \
            ", " + \
            repr(self.y) + \
            ")"